internal/util-logging/src/main/scala/sbt/internal/util/Terminal.scala

```diff
@@ -0,0 +1,210 @@
+/*
+ * sbt
+ * Copyright 2011 - 2018, Lightbend, Inc.
+ * Copyright 2008 - 2010, Mark Harrah
+ * Licensed under Apache License 2.0 (see LICENSE)
+ */
+
+package sbt.internal.util
+
+import java.io.{ InputStream, OutputStream }
+import java.util.Locale
+import java.util.concurrent.atomic.{ AtomicBoolean, AtomicReference }
+import java.util.concurrent.locks.ReentrantLock
+
+import jline.console.ConsoleReader
+
+import scala.util.control.NonFatal
+
+object Terminal {
+
+  /**
+   * Gets the current width of the terminal. The implementation reads a property from the jline
+   * config which is updated if it has been more than a second since the last update. It is thus
+   * possible for this value to be stale.
+   *
+   * @return the terminal width.
+   */
+  def getWidth: Int = terminal.getWidth
+
+  /**
+   * Gets the current height of the terminal. The implementation reads a property from the jline
+   * config which is updated if it has been more than a second since the last update. It is thus
+   * possible for this value to be stale.
+   *
+   * @return the terminal height.
+   */
+  def getHeight: Int = terminal.getHeight
+
+  /**
+   * Returns true if the current terminal supports ansi characters.
+   *
+   * @return true if the current terminal supports ansi escape codes.
+   */
+  def isAnsiSupported: Boolean =
+    try terminal.isAnsiSupported
+    catch { case NonFatal(_) => !isWindows }
+
+  /**
+   * Returns true if System.in is attached. When sbt is run as a subprocess, like in scripted or
+   * as a server, System.in will not be attached and this method will return false. Otherwise
+   * it will return true.
+   *
+   * @return true if System.in is attached.
+   */
+  def systemInIsAttached: Boolean = attached.get
+
+  /**
+   * Provides a wrapper around System.in. The wrapped stream in will check if the terminal is attached
+   * in available and read. If a read returns -1, it will mark System.in as unattached so that
+   * it can be detected by [[systemInIsAttached]].
+   *
+   * @return the wrapped InputStream
+   */
+  private[sbt] def wrappedSystemIn: InputStream = WrappedSystemIn
```

Comment:
Would it make sense to hide this from the rest of sbt code as well? And provide something like `def read: Char`?

Reply:
Possibly. `ConsoleChannel` uses `Terminal.throwOnClosedSystemIn` rather than wrapped system in but I think generally we want wrappedSystemIn.